docs/proposed/http-introducer-internals.rst

```diff
@@ -0,0 +1,192 @@
+.. -*- coding: utf-8 -*-
+
+HTTP Introducer Internals
+=========================
+
+This document explains the implementation of the "HTTP Introducer" system.
+
+Overview
+--------
+
+The HTTP Introducer is very similar to the Foolscap-based introducer it replaces.
+The primary difference is that it replaces the Foolscap-based protocol with a WebSocket-based protocol.
+In every other way,
+the HTTP Introducer is intended to replicate the behavior of the Foolscap-based introducer.
+
+http-introducer
+---------------
+
+The HTTP Introducer comprises a new command-line interface,
+a long-running process,
+and configuration for that process.
+
+
+Configuration
+-------------
+
+The command-line interface is configured with a JSON document.
+The JSON document follows this form::
+
+  { "version": 1
+  , "listen-endpoints": [ <strings> ]
+  , "network-location": [ <string> ]
+  , "certificate": <string>
+  , "private-key": <string>
+  , "swissnum": <string>
+  }
+
+The *version* property must be **1**.
+The values for the *listen-endpoints*, *certicate*, and *private-key* properties are those given to the ``create`` command.
+The *swissnum* property is string that is randomly generated at ``create``-time.
+
+Introducer fURL
+---------------
+
+.. TODO: What actual scheme will we use instead of "xxx"?
+.. TODO: Link to the discussion of security properties of this scheme in the GBS doc
+.. TODO: Update that doc to link to any Tor or Foolscap docs about the security properties of their systems
+
+The *introducer fURL* is derived from *certificate* and *swissnum*.
+It has the form ``xxx://<spki-hash>@<network-location>[,...]/<swissnum>``
+``<spki-hash>`` is the SPKI hash of the certificate.
+``<network-location>`` is one of the ``network-location`` property elements.
+Several locations may be present and separated by ``,``.
+``<swissnum>`` is the *swissnum* property value.
+
+The result is an unguessable self-authenticating URL which can be used to establish a confidential channel to the HTTP Introducer.
+
+Protocol
+--------
+
+The long-running process operates a TLS server on its configured endpoint.
+It uses the configured certificate and private key for the necessary TLS negotiation.
+On top of TLS,
+it runs a WebSocket server with two endpoints.
+
+The server does not require the client to present a certificate.
+The client requires the server to present a certificate with an SPKI hash matching that in the *introducer fURL*.
+
+.. TODO: Add docs about the WebSocket protocol negotiation that happens for the pub/sub protocol
```

Comment:
> I guess that leaves "what to do with the swissnum" since that's in the URIs below .. but so I guess that just means we just put "the one Introducer WebSocket endpoint at /<swissnum>" and do the duplix thing in your TODO below?

Yea I reckon so.  Adjusted the doc to reflect this.